lightning/src/ln/onion_route_tests.rs

```diff
@@ -352,143 +477,425 @@ fn test_onion_failure() {
+			);
+			let failure = onion_utils::build_failure_packet(
+				onion_keys[0].shared_secret.as_ref(),
+				LocalHTLCFailureReason::RequiredNodeFeature,
+				&[0; 0],
+				0,
+			);
+			msg.reason = failure.data;
+			msg.attribution_data = failure.attribution_data;
+		},
+		|| {
+			nodes[2].node.fail_htlc_backwards(&payment_hash);
+		},
+		true,
+		Some(LocalHTLCFailureReason::RequiredNodeFeature),
+		Some(NetworkUpdate::NodeFailure {
+			node_id: route.paths[0].hops[0].pubkey,
+			is_permanent: true,
+		}),
+		Some(route.paths[0].hops[0].short_channel_id),
+		Some(next_hop_failure.clone()),
+	);
 
 	// final node failure
-	run_onion_failure_test_with_fail_intercept("required_node_feature_missing", 200, &nodes, &route, &payment_hash, &payment_secret, |_msg| {}, |msg| {
-		let session_priv = SecretKey::from_slice(&[3; 32]).unwrap();
-		let onion_keys = onion_utils::construct_onion_keys(&Secp256k1::new(), &route.paths[0], &session_priv);
-		let failure = onion_utils::build_failure_packet(onion_keys[1].shared_secret.as_ref(), LocalHTLCFailureReason::RequiredNodeFeature, &[0;0], 0);
-		msg.reason = failure.data;
-		msg.attribution_data = failure.attribution_data;
-	}, ||{
-		nodes[2].node.fail_htlc_backwards(&payment_hash);
-	}, false, Some(LocalHTLCFailureReason::RequiredNodeFeature), Some(NetworkUpdate::NodeFailure{node_id: route.paths[0].hops[1].pubkey, is_permanent: true}), Some(route.paths[0].hops[1].short_channel_id), None);
+	run_onion_failure_test_with_fail_intercept(
+		"required_node_feature_missing",
+		200,
+		&nodes,
+		&route,
+		&payment_hash,
+		&payment_secret,
+		|_msg| {},
+		|msg| {
+			let session_priv = SecretKey::from_slice(&[3; 32]).unwrap();
+			let onion_keys = onion_utils::construct_onion_keys(
+				&Secp256k1::new(),
+				&route.paths[0],
+				&session_priv,
+			);
+			let failure = onion_utils::build_failure_packet(
+				onion_keys[1].shared_secret.as_ref(),
+				LocalHTLCFailureReason::RequiredNodeFeature,
+				&[0; 0],
+				0,
+			);
+			msg.reason = failure.data;
+			msg.attribution_data = failure.attribution_data;
+		},
+		|| {
+			nodes[2].node.fail_htlc_backwards(&payment_hash);
+		},
+		false,
+		Some(LocalHTLCFailureReason::RequiredNodeFeature),
+		Some(NetworkUpdate::NodeFailure {
+			node_id: route.paths[0].hops[1].pubkey,
+			is_permanent: true,
+		}),
+		Some(route.paths[0].hops[1].short_channel_id),
+		None,
+	);
 	let (_, payment_hash, payment_secret) = get_payment_preimage_hash!(nodes[2]);
 
 	// Our immediate peer sent UpdateFailMalformedHTLC because it couldn't understand the onion in
 	// the UpdateAddHTLC that we sent.
 	let short_channel_id = channels[0].0.contents.short_channel_id;
-	run_onion_failure_test("invalid_onion_version", 0, &nodes, &route, &payment_hash, &payment_secret, |msg| { msg.onion_routing_packet.version = 1; }, ||{}, true,
-		Some(LocalHTLCFailureReason::InvalidOnionVersion), None, Some(short_channel_id), Some(HTLCHandlingFailureType::InvalidOnion));
+	run_onion_failure_test(
```

Comment:
It is vertical, but I still like it in my IDE with types:

![image](https://github.com/user-attachments/assets/f347fa3a-a362-4363-b512-9066b9001c39)